pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -169,8 +174,91 @@ pub mod pallet {
 	impl<T: Config> Pallet<T> {
 		#[pallet::call_index(0)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn create_pool(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			curve: CurveInput<CurveParameterInputOf<T>>,
+			currencies: BoundedVec<TokenMetaOf<T>, T::MaxCurrencies>,
+			state: PoolStatus<Locks>,
+			denomination: u8,
+			pool_manager: AccountIdOf<T>,
+			transferable: bool,
+			team: Team<AccountIdOf<T>>,
+		) -> DispatchResult {
+			let who = T::PoolCreateOrigin::ensure_origin(origin)?;
+
+			let currency_length = currencies.len();
+
+			ensure!(
+				(1..=(T::MaxCurrencies::get()).saturated_into()).contains(&currency_length),
+				Error::<T>::CurrenciesNumber
+			);
+
+			let checked_curve = curve.try_into().map_err(|_| Error::<T>::InvalidCoefficients)?;
+
+			let current_asset_id = NextAssetId::<T>::get();
+
+			let (currency_ids, next_asset_id) = Self::generate_sequential_asset_ids(current_asset_id, currency_length)?;
+
+			let pool_id = T::PoolId::from(currency_ids.blake2_256());
+
+			// Todo: change that.
+			T::DepositCurrency::hold(
+				&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				&who,
+				T::BaseDeposit::get().saturating_add(
+					T::DepositPerCurrency::get()
+						.saturating_mul(currency_length.saturated_into())
+						.saturated_into(),
+				),
+			)?;
+
+			currencies
+				.iter()
+				.enumerate()
+				.try_for_each(|(idx, entry)| -> DispatchResult {
```

Comment:
Move this closure to its own function. Benchmarking would be simpler.

Reply:
Still, good luck benchmarking this 😃 Since as far as I understand it's not a requirement that all currencies have sequential IDs, have you considered exposing one more extrinsic `add_currency` which would do one iteration of this foor loop, and a `finalize_curve` or something like that which prevents new currency additions?

Reply:
The power of tx batching is real.

Reply:
Of course, as long as the `create_pool` slims down to just setting the basic curve parameters like denomination and admin team.

Reply:
Is it that difficult to benchmark? I would benchmark the closure separately and the transaction without any currencies. The final weight would be the length of the currencies * the weight from the closure + the base weight from the transaction. That should be pretty accurate.

Reply:
Writing smaller atomic tx is ofc another option, but I would say it is not as complicated yet to do that

Reply:
The way you would benchmark it is actually just giving it input currencies from 1 to the max limit, that's it. The implementation of the `Fungibles` for the pallet should be part of the benchmarked logic, so I don't think you need to split it up and benchmark the closure separately.

Reply:
So yeah, for input vecs probably it's ok to benchmark it like that, but for other extrinsics which iterate over stored vecs or other collections, it's better to expose them via atomic transactions that operate directly on a single item.

Reply:
So yeah, benchmarking this might actually not be that hard, although I would still suggest to take out of the extrinsic signature stuff that could be done in its own extrinsic, like setting the team, for instance. Then, I would also not split this up at all, but simply benchmark the whole extrinsic based on the length of the input vec, nothing else.

Reply:
I need to investigate that. I can imagine, benchmarking the worst case can exceed the resources of a block, depending on how many currencies we might allow. For now, I moved out the reset of the team to its own tx: [0f8c286](https://github.com/KILTprotocol/kilt-node/pull/768/commits/0f8c286505b02e167123d6d89164437d34a17fe9)

Reply:
Yes but if we calculate weights based on the length of the input parameter, and that would make a tx too big, then the tx is ACTUALLY too big. That's not a worst case weight, that is the actual weight the tx takes, since we take the exact size of the input in consideration.